zerver/tests/test_message_fetch.py

```diff
@@ -2085,6 +2085,97 @@ def test_unauthenticated_get_messages(self) -> None:
+        messages = orjson.loads(result.content)["messages"]
+        self.assert_length(messages, 2)
+        fetched_message_ids = [message["id"] for message in messages]
+        self.assertEqual(fetched_message_ids.sort(), [msg1, msg2].sort())
+
+    def test_get_message_ids_web_public(self) -> None:
+        self.login("iago")
+        self.subscribe(self.example_user("iago"), "Rome")
+        self.logout()
+        msg1 = self.send_stream_message(self.example_user("iago"), "Rome")
+        msg2 = self.send_stream_message(self.example_user("iago"), "Rome")
+        result = self.client_get(
+            "/json/messages",
+            {
+                "message_ids": orjson.dumps([msg1, msg2]).decode(),
+                "narrow": orjson.dumps([dict(operator="channels", operand="web-public")]).decode(),
+            },
+        )
+
+        self.assert_json_success(result)
+        messages = orjson.loads(result.content)["messages"]
+        self.assert_length(messages, 2)
+        fetched_message_ids = [message["id"] for message in messages]
+        self.assertEqual(fetched_message_ids.sort(), [msg1, msg2].sort())
+
+    def test_message_fetch_with_mutually_exclusive_parameters(self) -> None:
+        mutually_exclusive_params_with_message_ids = ["num_before", "num_after", "anchor"]
+        for param in mutually_exclusive_params_with_message_ids:
+            result = self.client_get(
+                "/json/messages",
+                {
+                    "message_ids": orjson.dumps([1, 2]).decode(),
+                    param: 1,
+                },
+            )
+            error_msg = "Unsupported parameter combination: num_before, num_after, anchor, message_ids, use_first_unread_anchor_val"
+            self.assert_json_error(result, error_msg)
+
+    def test_message_fetch_for_inaccessible_message_ids(self) -> None:
+        # Add new channels
+        realm = get_realm("zulip")
+        channel_dicts: list[StreamDict] = [
+            {
+                "name": "private-channel",
+                "description": "Private channel with non-public history",
+                "invite_only": True,
+            },
+            {
+                "name": "private-channel-with-history",
+                "description": "Private channel with public history",
+                "invite_only": True,
+                "history_public_to_subscribers": True,
+            },
+        ]
+        create_streams_if_needed(realm, channel_dicts)
+
+        # Unsubscribe `hamlet` (our test user) from all 3 streams.
+        hamlet = self.example_user("hamlet")
+        iago = self.example_user("iago")
+        self.login("iago")
+        message_ids = []
+        for stream_name in ["private-channel", "private-channel-with-history"]:
+            self.unsubscribe(hamlet, stream_name)
+            self.subscribe(iago, stream_name)
+            message_ids.append(self.send_stream_message(self.example_user("iago"), stream_name))
+        self.logout()
```

Comment:
This seems a bit contorted, switching user accounts so many times. Generally we use `self.api_post` to avoid needing to switch users for individual operations ... but also I don't see how `hamlet` would have been subscribed to those channels in the first place, if they're newly created by Iago?

Reply:
Okay, removed the hamlet calls to unsubscribe here. I thought the test might be more idempotent with those calls.

Reply:
Tests don't need to be idempotent from a database state standpoint; in Django, the test runs in a transaction that is rolled back at the end of the test.